File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (root == NULL) {
            return res;
        }

        deque<TreeNode*> queue;

        queue.push_back(root);

        while (!queue.empty()) {
            vector<int> list;
            int size = queue.size();

            while (size > 0) {
                TreeNode* node = queue.front();
                queue.pop_front();

                if (node->left != NULL) {
                    queue.push_back(node->left);
                }

                if (node->right != NULL) {
                    queue.push_back(node->right);
                }

                list.push_back(node->val);
                size--;
            }
            res.push_back(list);
        }

        return res;
    }
};

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []

        if not root:
            return res

        # queue = deque([root])

        # while queue:
        #     size = len(queue)
        #     level = []

        #     for _ in range(size):
        #         node = queue.popleft()

        #         level.append(node.val)

        #         if node.left:
        #             queue.append(node.left)

        #         if node.right:
        #             queue.append(node.right)
        #     res.append(level)

        def traverse(curr, level):
            if not curr:
                return
            
            if len(res) == level:
                res.append([])

            res[level].append(curr.val)

            traverse(curr.left, level + 1)
            traverse(curr.right, level + 1)


        traverse(root, 0)

        return res 